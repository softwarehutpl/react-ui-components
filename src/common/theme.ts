const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
