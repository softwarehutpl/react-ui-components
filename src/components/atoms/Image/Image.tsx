import React, { useEffect } from 'react';
import styled from 'styled-components';
import { OBJECT_FIT_OPTIONS } from '../../../common/constants/consts';

interface IImage {
  className?: string;
  src: string;
  width?: number;
  height?: number;
  objectFit?: 'cover' | 'contain' | 'fill' | 'none' | 'scale-down';
  alt: string;
  lazy?: boolean;
  rootMargin?: string;
  threshold?: number;
}

const Image = ({ className, src, alt, lazy, rootMargin, threshold }: IImage) => {
  const imageRef = React.useRef<HTMLImageElement>(null);

  const preload = (image: any, imageSrc: string) => {
    image.src = imageSrc;
  };

  const observer = new IntersectionObserver(
    (entries, self) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          preload(entry.target, src);
          self.unobserve(entry.target);
        }
      });
    },
    { rootMargin, threshold }
  );

  useEffect(() => {
    if (imageRef.current && lazy) {
      const image = imageRef.current;
      observer.observe(image);
    }
  }, []);

  return <img ref={imageRef} className={className} alt={alt} src={lazy ? '' : src} />;
};

const defaultProps = {
  objectFit: OBJECT_FIT_OPTIONS.cover,
  lazy: false,
  rootMargin: '0px',
  threshold: 0,
};

Image.defaultProps = defaultProps;

export default styled(Image)<IImage>`
  ${({ objectFit = defaultProps.objectFit, width, height }) => ({
    'object-fit': objectFit,
    width: width || '100%',
    height: height || '100%',
  })}
`;
