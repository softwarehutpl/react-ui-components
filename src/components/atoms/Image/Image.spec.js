import React from 'react';
import { create } from 'react-test-renderer';
import 'jest-styled-components';
import theme from '../../../common/theme';
import Image from '../../atoms/Image/Image';
import setupIntersectionObserverMock from '../../../helpers/intersectionObserverMock';

describe('Image component', () => {
  beforeEach(() => {
    setupIntersectionObserverMock();
  });

  it('should render correctly', () => {
    const image = create(
      <Image theme={theme} src="https://via.placeholder.com/150" alt="placeholder" />
    ).toJSON();
    expect(image).toMatchSnapshot();
  });

  it('should render image with custom className', () => {
    const image = create(<Image theme={theme} className="test-class" />).toJSON();
    expect(image.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render image with given size', () => {
    const image = create(<Image theme={theme} width={200} height={400} />).toJSON();
    expect(image).toHaveStyleRule('width', '200px');
    expect(image).toHaveStyleRule('height', '400px');
  });

  it('should render image with given src and alt properties', () => {
    const image = create(
      <Image theme={theme} src="https://via.placeholder.com/150" alt="placeholder" />
    ).toJSON();
    expect(image.props.src).toBe('https://via.placeholder.com/150');
    expect(image.props.alt).toBe('placeholder');
  });

  it('should render image with given object-fit', () => {
    const image = create(<Image theme={theme} objectFit="contain" />).toJSON();
    expect(image).toHaveStyleRule('object-fit', 'contain');
  });
});
