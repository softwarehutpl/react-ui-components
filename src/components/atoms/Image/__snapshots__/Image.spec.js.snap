// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component should render correctly 1`] = `
.c0 {
  object-fit: cover;
  width: 100%;
  height: 100%;
}

<img
  alt="placeholder"
  className="c0"
  src="https://via.placeholder.com/150"
/>
`;
